flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1420,6 +1424,41 @@ private void timeoutSlot(AllocationID allocationId, UUID ticket) {
 		}
 	}
 
+	/**
+	 * Sync slot with job master.
+	 * The slots only exist in job master will be failed to job master.
+	 * The slots only exist in task executor will be released to resource manager.
+	 *
+	 * @param slotSnapshot is the snapshot of slots in job master
+	 */
+	private void syncSlotWithSnapshotFromJobMaster(SlotSnapshot slotSnapshot) {
+		JobManagerConnection jobManagerConnection = jobManagerTable.get(slotSnapshot.getJobId());
+		if (jobManagerConnection != null) {
+			JobMasterGateway jobMasterGateway = jobManagerConnection.getJobManagerGateway();
+
+			for (OfferedSlot offeredSlot : slotSnapshot.getSlots()) {
+				AllocationID allocationId = offeredSlot.getAllocationId();
+				if (!taskSlotTable.isAllocated(offeredSlot.getSlotIndex(), slotSnapshot.getJobId(), allocationId)) {
+					jobMasterGateway.failSlot(
+							getResourceID(),
+							allocationId,
+							new Exception("Slot " + offeredSlot.getSlotIndex() + " is not allocated by " + allocationId));
+				}
+			}
+		}
+
+		Set<AllocationID> diffWithJobMaster = new HashSet<>();
+		Iterator<AllocationID> slotsTaskManagerSide = taskSlotTable.getActiveSlots(slotSnapshot.getJobId());
+		while (slotsTaskManagerSide.hasNext()) {
+			AllocationID allocationId = slotsTaskManagerSide.next();
+			if (!slotSnapshot.contains(allocationId)) {
+				diffWithJobMaster.add(allocationId);
+			}
+		}
+		diffWithJobMaster.forEach(
+				(allocationId) -> freeSlotInternal(allocationId, new Exception(allocationId + " not exists in job master.")));
```

Comment:
We could streamline this last part a bit by using existing functionality:
```
private void syncSlotsWithSnapshotFromJobMaster(SlotSnapshot slotSnapshot) {
		final JobManagerConnection jobManagerConnection = jobManagerTable.get(slotSnapshot.getJobId());
		if (jobManagerConnection != null) {
			final JobMasterGateway jobMasterGateway = jobManagerConnection.getJobManagerGateway();

			failNoLongerAllocatedSlots(slotSnapshot, jobMasterGateway);

			freeNoLongerUsedSlots(slotSnapshot);
		} else {
			log.debug("Ignoring allocated slot report from job " + slotSnapshot.getJobId() + " because there is no active leader.");
		}
	}

	private void failNoLongerAllocatedSlots(SlotSnapshot slotSnapshot, JobMasterGateway jobMasterGateway) {
		for (OfferedSlot offeredSlot : slotSnapshot.getSlots()) {
			final AllocationID allocationId = offeredSlot.getAllocationId();
			if (!taskSlotTable.isAllocated(offeredSlot.getSlotIndex(), slotSnapshot.getJobId(), allocationId)) {
				jobMasterGateway.failSlot(
					getResourceID(),
					allocationId,
					new FlinkException("Slot " + offeredSlot.getSlotIndex() + " on TaskExecutor " + getResourceID() + " is not allocated by job " + slotSnapshot.getJobId() + '.'));
			}
		}
	}

	private void freeNoLongerUsedSlots(SlotSnapshot slotSnapshot) {
		final Iterator<AllocationID> slotsTaskManagerSide = taskSlotTable.getActiveSlots(slotSnapshot.getJobId());
		final Set<AllocationID> activeSlots = Sets.newHashSet(slotsTaskManagerSide);
		final Set<AllocationID> reportedSlots = slotSnapshot.getSlots().stream().map(OfferedSlot::getAllocationId).collect(Collectors.toSet());

		final Sets.SetView<AllocationID> difference = Sets.difference(activeSlots, reportedSlots);

		for (AllocationID allocationID : difference) {
			freeSlotInternal(allocationID, new FlinkException(allocationID + " is no longer allocated by job " + slotSnapshot.getJobId()));
		}
	}
```

Reply:
Good point